python_on_whales/components/container/cli_wrapper.py

```diff
@@ -704,6 +702,8 @@ def create(
         self._add_publish_to_command(full_cmd, publish)
         full_cmd.add_flag("--publish-all", publish_all)
 
+        full_cmd.add_simple_arg("--pull", pull)
```

Comment:
the pull was done separately before so that it doesn't appear in stderr and stdout when getting the output of the container. I think we should keep it that way :)

Reply:
Hi :)

If I understood correctly, the "pull" you are referring to is done by the function `_pull_if_necessary` called inside the bodies of `run` and `create` right? If so, this function _forces_ the behavior of `"missing"` (an image is always pulled if it does not exist locally). It does not support the `"always"` and `"never"` behaviors in any way (hence the open issue).

I had not considered the _stderr_ and _stdout_ related issue (whoops) and I totally agree with you.

Having all this into consideration I've made new changes to the code.
Now, the same new features I described earlier are kept and no image pull message appears in the container  _stream_ (check the following example using `"always"`).

```python
>>> from python_on_whales import docker
>>> stream = docker.run('alpine:latest', command=['ls'], stream=True, pull='always')
latest: Pulling from library/alpine
Digest: sha256:bc41182d7ef5ffc53a40b044e725193bc10142a1243f395ee852a8d9730fc2ad
Status: Image is up to date for alpine:latest
>>> for msg in stream:
...     print(msg)
... 
('stdout', b'bin\n')
('stdout', b'dev\n')
('stdout', b'etc\n')
('stdout', b'home\n')
('stdout', b'lib\n')
('stdout', b'media\n')
('stdout', b'mnt\n')
('stdout', b'opt\n')
('stdout', b'proc\n')
('stdout', b'root\n')
('stdout', b'run\n')
('stdout', b'sbin\n')
('stdout', b'srv\n')
('stdout', b'sys\n')
('stdout', b'tmp\n')
('stdout', b'usr\n')
('stdout', b'var\n')
```

New changes also include updated unit tests for the `run` and `create` functions.

Reply:
Very nice analysis thanks :)